lib: os: cbprintf: Add support for conversion to fully self-contained package

nordic-krch opened the pull request:
Added support for conversion from a standard package which contains
pointers to read only strings to fully self-contained (fsc) package.
Fsc package contains all strings associated with the package thus
access to read only strings is not needed to format a string.
    
In order to allow conversion to fsc package, standard package must
contain locations of all string pointers within the package. Appending
that information is optional and is controlled by flags parameter
which was added to packaging API. If option flag is set then
package contains header, arguments, locations of read only strings and
transient strings (each prefixed with string argument location).
Package header has been extended with field which contains number of
read only string locations.
    
A function for conversion to fsc package has been added
(cbprintf_fsc_package()).

Use case
=======
Logging will be extended to support multiple domains. In that scenario, messages will be passed between domains and context which will be finally formatting the string will have no access to read only address from remote domain. However, static package creation is fast and it is still a must in logging context. This means that, before log message (essentially, a package) is passed to another domain, it must be converted to a format which will contain also read only strings (`fully self-contained package`). Conversion is straightforward as long as we know where string pointers are located in the package. For now, package contained that information only for strings which were copied into package body. This PR adds option to append that information also for read only strings. Appending this data shall be optional since it is needed only in scenario where conversion for fsc package may happen.

Since when domains are crossed alignment may change, i will also be looking into packed version of a package. That will come in near future.

Opening as draft for now, need to extend `tests/lib/cbprintf_packge` suite as well.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
lib: os: cbprintf: Add support for conversion to fsc package

Added support for conversion from a standard package which contains
pointers to read only strings to fully self-contained (fsc) package.
Fsc package contains all strings associated with the package thus
access to read only strings is not needed to format a string.

In order to allow conversion to fsc package, standard package must
contain locations of all string pointers within the package. Appending
that information is optional and is controlled by flags parameter
which was added to packaging API. If option flag is set then
package contains header, arguments, locations of read only strings and
transient strings (each prefixed with string argument location).
Package header has been extended with field which contains number of
read only string locations.

A function for conversion to fsc package has been added
(cbprintf_fsc_package()).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: unit: cbprintf: Add test for fsc package

Added test cases for cbprintf_fsc_package() and use of
CBPRINTF_PACKAGE_ADD_STRING_IDXS flag.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

npitre reviewed (approved):
This assumes  will always be given a size exactly where the package ends.

npitre reviewed (commented):
This assumes  will always be given a size exactly where the package ends.

nordic-krch committed:
tests: lib: cbprintf: Add tests for fsc package

Added tests for fully self-contained packages and use of
CBPRINTF_PACKAGE_ADD_STRING_IDXS flag.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
Weird compliance failure about leading spaces in doxygen comment:
```
-:114: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#114: FILE: include/sys/cbprintf.h:177:
+ * @param flags option flags. See @ref CBPRINTF_PACKAGE_FLAGS.$

-:114: ERROR:SPACING: space prohibited after that '*' (ctx:ExW)
#114: FILE: include/sys/cbprintf.h:177:
+ * @param flags option flags. See @ref CBPRINTF_PACKAGE_FLAGS.
  ^

-:115: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#115: FILE: include/sys/cbprintf.h:178:
+ *$
```
@nashif any idea?
